### PR TITLE
[DRAFT] First version of fusion optimizations for transformers

### PR DESCRIPTION
**Pull request by gramalingam**
* Introduce fusion rules for SdpaAttention, RMS Normalization, Skip Normalization, Rotary Embedding, and Multi Head Attention
* Replace Expand by Identity when applicable (in core optimization)
* Cleanup Dropout Identity replacement in the case when Dropout has mask output
* Make repeated (redundant) call to inliner efficient

Still TODO: 
* Multi Head Attention requires extra validation conditions
* Need to cleanup use of "local" sub-patterns

### CODE REVIEW

File: onnxscript/optimizer/_constant_folding.py [hunk 1]
```diff
@@ -434,25 +435,54 @@ def concat(node: ir.Node, op, state: OptimizerState) -> ReturnValue:
...
-        # TODO: handle this case. But unlikely to be needed in practice.
-        return None
+
+    def optimized_dropout():
+        input = node.inputs[0]
+        output = op.Identity(input)
+        if len(node.outputs) == 1:
+            return output
+        else:
+            true_tensor = onnx.helper.make_tensor("true", onnx.TensorProto.BOOL, [1], [True])
```
Comment (justinchuby) [inline onnxscript/optimizer/_constant_folding.py:445]: Is this IR? If so
```suggestion
            true_tensor = ir.tensor([True])
```

Comment (gramalingam) [reply onnxscript/optimizer/_constant_folding.py:445]: Thanks. But when I look at the signature or code [here](https://github.com/microsoft/onnxscript/blob/32090a8d635b297cfd90032c00b39777415d4054/onnxscript/ir/_convenience.py#L294), it is not clear this is supported or will work?

Comment (justinchuby) [reply onnxscript/optimizer/_constant_folding.py:445]: Good catch. We can update the signature

Comment (justinchuby) [reply onnxscript/optimizer/_constant_folding.py:459]: This is actually covered by npt.ArrayLike

Comment (gramalingam) [reply onnxscript/optimizer/_constant_folding.py:459]: Actually, I tried and it failed, rejecting a list. BTW, I have moved the independent parts of this PR into a separate PR: https://github.com/microsoft/onnxscript/pull/1947

Comment (justinchuby) [reply onnxscript/optimizer/_constant_folding.py:459]: Hmm. I need to fix that then

Comment (justinchuby) [reply onnxscript/optimizer/_constant_folding.py]: Resolved

**Comment by justinchuby** [comment]
I created
- https://github.com/microsoft/onnxscript/pull/1952

**Comment by gramalingam** [comment]
Closing this for now. Will recreate as multiple separate PRs.

**Pull request closed without merging**